### ISSUE TITLE
GraalVM generates wrong result due to speculative optimization

### ISSUE THREAD

**Issue by helloguo** [oracle/graal#2493]
**Describe the issue**
GraalVM generates wrong result due to speculative optimization. 

**Steps to reproduce the issue**
Please include both build steps as well as run steps
1. Run this code
If you cannot reproduce it, you can simply increase the `int size = 1000000` to a lager value in `main()` function to make the repro work.
```
import java.util.HashMap;

class InputRow {
  public long[] row;
  public int len;

  public InputRow(int size, long initValue) {
    len = size;
    row = new long[size];
    for (int i = 0; i < size; i++) {
      row[i] = initValue;
    }
  }

  public long getLong(int idx) {
    assert (idx < len);
    return row[idx];
  }
}

class SimpleInput {
  public InputRow[] rows;
  public int len;
  public int idx;

  public SimpleInput(int size) {
    len = size;
    rows = new InputRow[size];
    idx = 0;
    for (int i = 0; i < size - 3; i = i + 3) {
      rows[i] = new InputRow(10, 50);
      rows[i + 1] = new InputRow(10, -1);
      rows[i + 2] = new InputRow(10, 100);
    }
    rows[size - 1] = new InputRow(10, 10);
  }

  public boolean hasNext() {
    return idx < len;
  }

  public InputRow next() {
    InputRow ret = rows[idx];
    idx++;
    return ret;
  }
}

class OutputElement {
  public long value1;
  public String value2;

  public OutputElement() {
    value1 = 0;
    value2 = "init";
  }

  public void write(int idx, long value) {
    assert (idx == 0);
    value1 = value;
  }

  public void write(int idx, String value) {
    assert (idx == 1);
    value2 = value;
  }
}

class SimpleOutput {
  public OutputElement[] output;
  public int len;
  public int idx;

  public SimpleOutput(int size) {
    output = new OutputElement[size];
    len = size;
    idx = 0;
  }

  public void append(OutputElement element) {
    assert (idx < len);
    output[idx] = element;
    idx++;
  }

  public void print() {
    HashMap<String, Integer> map = new HashMap<String, Integer>();
    for (int i = 0; i < len; i++) {
      long v1 = output[i].value1;
      String v2 = output[i].value2;
      if (map.containsKey(v2)) {
        int tmp = map.get(v2);
        map.put(v2, (tmp + 1));
      } else {
        map.put(v2, 1);
      }
    }
    System.out.print(map);
  }
}

public class Main {
  static void processNext(SimpleInput input, SimpleOutput output) {
    while (input.hasNext()) {
      InputRow orc_scan_row = (InputRow) input.next();
      long orc_scan_value = orc_scan_row.getLong(0);
      String project_value1 = null;
      boolean project_value2 = false;
      if (orc_scan_value >= 0L) {
        project_value2 = orc_scan_value <= 35L;
      }
      if (project_value2) {
        project_value1 = "35";
      } else {
        if (orc_scan_value > 35L) {
          project_value1 = "90";
        } else {
          project_value1 = "Unknown";
        }
      }
      OutputElement element = new OutputElement();
      element.write(0, orc_scan_value);
      element.write(1, project_value1);
      output.append(element);
    }
  }

  public static void main(String[] args) {
    System.out.println("Hello World!");
    int size = 1000000;
    SimpleInput input = new SimpleInput(size);
    SimpleOutput output = new SimpleOutput(size);
    processNext(input, output);
    output.print();
  }
}
```
2. Output of GraalVM
```
Hello World!
{Unknown=333334, 90=666666}
```
3. Output of OpenJDK
```
Hello World!
{35=1, Unknown=333333, 90=666666}
```
**Describe GraalVM and your environment:**
 - GraalVM version (latest snapshot builds can be found [here](https://github.com/graalvm/graalvm-ce-dev-builds/releases)), or commit id if built from source: **graalvm-ce-java8-19.3.1**
 - JDK major version: **[8]**
 - OS: **[macOS Catalina]**
 - Architecture: **[AMD64]**

**More details**
-Dgraal.RemoveNeverExecutedCode=false could generate the correct result.

**Comment by helloguo** [comment]
cc @christhalinger

**Comment by helloguo** [comment]
The bug could be reproduced on CentOS 7 as well.

**Comment by helloguo** [comment]
I was trying to dump the graph to have better understanding of the issue. However, the result was different when I dumped the graph, as shown below:
```
helloguo-mbp:bug helloguo$ ./../Desktop/graalvm-ce-java8-19.3.1/Contents/Home/bin/java  Main
Hello World!
{Unknown=333334, 90=666666}helloguo-mbp:bug helloguo$ 
helloguo-mbp:bug helloguo$ 
helloguo-mbp:bug helloguo$ 
helloguo-mbp:bug helloguo$ ./../Desktop/graalvm-ce-java8-19.3.1/Contents/Home/bin/java -Dgraal.Dump  -Dgraal.MethodFilter=*.processNext Main
Hello World!
[Use -Dgraal.LogFile=<path> to redirect Graal log output to a file.]
Dumping IGV graphs in /Users/helloguo/bug/graal_dumps/2020.05.22.21.31.32.151
{35=1, Unknown=333333, 90=666666}helloguo-mbp:bug helloguo$ 
```

**Comment by dougxc** [comment]
Thanks for the report @helloguo . This looks like a bad compilation of `Main.processNext` since the bad result is always produced by `java -Dgraal.GraalCompileOnly=processNext Main`. We will investigate.

**Comment by helloguo** [comment]
@dougxc @davleopo Thanks for looking at this issue.

One thing may worth mentioning is that the bug cannot be reproduced when I add `-Dgraal.Dump  -Dgraal.MethodFilter=*.processNext ` to dump the graphs. In this case, I have to increase `int size = 1000000` to a lager value (e.g. 10000000) in `main()` function to make the reproduce case work. Now I have two sets of graphs: one is from the run which generates the correct result while the other one is from the run which generates the wrong result. When I compare the two sets of graphs, I'm not able to find the difference with IdealGraphVisualizer. Maybe I missed something there. Anyway, just share it here if it helps the investigation a bit.

While we are trying to narrow down the root cause, do you think `-Dgraal.RemoveNeverExecutedCode=false` is a proper short-term fix? The bug disappeared when I disabled RemoveNeverExecutedCode.

**Comment by dougxc** [comment]
I've investigated a bit more and suspect it's a problem in the code that [converts a signed range test into an unsigned compare](https://github.com/oracle/graal/commit/23b59b0f3aaa2343a6f935e18cf58384965c7348) but am not yet 100% sure.
In any case, I can reliably reproduce now with:
```
java -XX:+UseJVMCICompiler -XX:-TieredCompilation -Dgraal.GraalCompileOnly=processNext -Xbatch Main
```

**Comment by dougxc** [comment]
BTW, I don't think `-Dgraal.RemoveNeverExecutedCode=false` is a reliable workaround.

**Comment by lic9** [comment]
The problem seems relate to "[reordering two if statements](https://github.com/oracle/graal/blob/master/compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/IfNode.java#L304)" 

Before CanonicalizerPhase:
![Screen Shot 2020-05-23 at 7 28 40 PM](https://user-images.githubusercontent.com/1966176/82747058-8869d780-9d4a-11ea-89c8-337d8e167f7d.png)

"checkForUnsignedCompare()"'s generated new unsigned-compare "382 if" node looks correct, but reordering put "48 if" node before "382 if" so "382 if"'s trueSuccessor could never be reached.

![Screen Shot 2020-05-23 at 7 29 07 PM](https://user-images.githubusercontent.com/1966176/82747096-e4346080-9d4a-11ea-83ea-63deb310ead7.png)

Comment out line [304 - 338](https://github.com/oracle/graal/blob/master/compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/IfNode.java#L304) will get the correct result.

**Comment by plokhotnyuk** [comment]
@dougxc Please also consider revising of "merging range compare into an unsigned compare" to work on all ranges circa `C1 < x && x < C2` where `C2 - C1` does not overflow.

Below are code generated for a simple range check `b >= '0' && b <= '9'`.

OpenJDK 15:
```
[info]   0.31%    │  │ │ ││││   0x00007f3e580036af:   mov    %r8d,%ebx
[info]   0.29%    │  │ │ ││││   0x00007f3e580036b2:   add    $0xffffffd0,%ebx
[info]   0.37%    │  │ │ ││││   0x00007f3e580036b5:   cmp    $0xa,%ebx
[info]            │  ╰ │ ││││   0x00007f3e580036b8:   jae    0x00007f3e580034c3           ;*if_icmpgt {reexecute=0 rethrow=0 return_oop=0}
[info]            │    │ ││││                                                             ; - com.github.plokhotnyuk.jsoniter_scala.core.JsonReader::parseLong@165 (line 1318)
[info]            │    │ ││││                                                             ; - com.github.plokhotnyuk.jsoniter_scala.core.JsonReader::readLong@2 (line 376)
[info]            │    │ ││││                                                             ; - com.github.plokhotnyuk.jsoniter_scala.benchmark.JsoniterScalaCodecs$$anon$27::d0@57 (line 63)
```

GraalVM EE:
```
[info]   2.46%  │││││  │ ││    0x00007f039833640f: cmp    $0x30,%eax
[info]          │││││╭ │ ││    0x00007f0398336412: jl     0x00007f0398336473
[info]   3.68%  ││││││ │ ││    0x00007f0398336418: nopl   0x0(%rax,%rax,1)
[info]   2.50%  ││││││ │ ││    0x00007f0398336420: cmp    $0x3a,%eax
[info]          ││││││╭│ ││    0x00007f0398336423: jae    0x00007f0398336473  ;*ifeq {reexecute=0 rethrow=0 return_oop=0}
[info]          ││││││││ ││                                                  ; - com.github.plokhotnyuk.jsoniter_scala.core.JsonReader::parseLong@173 (line 1318)
[info]          ││││││││ ││                                                  ; - com.github.plokhotnyuk.jsoniter_scala.core.JsonReader::readLong@2 (line 376)
[info]          ││││││││ ││                                                  ; - com.github.plokhotnyuk.jsoniter_scala.benchmark.JsoniterScalaCodecs$$anon$27::d0@57 (line 63)
[info]          ││││││││ ││                                                  ; - com.github.plokhotnyuk.jsoniter_scala.benchmark.JsoniterScalaCodecs$$anon$27::decodeValue@3 (line 63)
[info]          ││││││││ ││                                                  ; - com.github.plokhotnyuk.jsoniter_scala.benchmark.JsoniterScalaCodecs$$anon$27::decodeValue@6 (line 63)
[info]          ││││││││ ││                                                  ; - com.github.plokhotnyuk.jsoniter_scala.core.JsonReader::read@46 (line 631)
[info]          ││││││││ ││                                                  ; - com.github.plokhotnyuk.jsoniter_scala.core.package$::readFromArray@15 (line 100)
[info]          ││││││││ ││                                                  ; - com.github.plokhotnyuk.jsoniter_scala.benchmark.ArrayOfLongsReading::jsoniterScala@19 (line 36)
[info]          ││││││││ ││                                                  ; - com.github.plokhotnyuk.jsoniter_scala.benchmark.generated.ArrayOfLongsReading_jsoniterScala_jmhTest::jsoniterScala_thrpt_jmhStub@1
```

GraalVM CE:
```
[info]   2.78%  │││││  │ │ │  │  0x00007f96a82cab1f: nop
[info]   2.60%  │││││  │ │ │  │  0x00007f96a82cab20: cmp    $0x30,%r13d
[info]          │││││╭ │ │ │  │  0x00007f96a82cab24: jl     0x00007f96a82cab8a
[info]   3.09%  ││││││ │ │ │  │  0x00007f96a82cab2a: cmp    $0x3a,%r13d
[info]          ││││││╭│ │ │  │  0x00007f96a82cab2e: jae    0x00007f96a82cab8a  ;*ifeq {reexecute=0 rethrow=0 return_oop=0}
[info]          ││││││││ │ │  │                                                ; - com.github.plokhotnyuk.jsoniter_scala.core.JsonReader::parseLong@173 (line 1318)
[info]          ││││││││ │ │  │                                                ; - com.github.plokhotnyuk.jsoniter_scala.core.JsonReader::readLong@2 (line 376)
[info]          ││││││││ │ │  │                                                ; - com.github.plokhotnyuk.jsoniter_scala.benchmark.JsoniterScalaCodecs$$anon$27::d0@57 (line 63)
```

**Comment by lic9** [comment]
It seems like [prepareForSwap(tool, condition(), nextIf.condition())](https://github.com/oracle/graal/blob/master/compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/IfNode.java#L312) doesn't handle signed comparison and unsigned comparison correctly. 

```
compareA is node "381||<|": condition=BT, X=27|Invoke#Direct#InputRow.getLong, Y=38|Constant(36, i64)
compareB is node "39|<": condition=LT, X=27|Invoke#Direct#InputRow.getLong, Y=38|Constant(36, i64)
Can swap disjoint coditions on same values: BT and LT
```

But they are not disjoint condition because conditionA is unsigned

**Comment by plokhotnyuk** [comment]
@dougxc Please, also, revise merging of range check compare operations to one for arbitrary range constants.

As example `b >= '0' && b <= '9'` can be replaced by `b + 2147483600 <= -2147483639`.

Please, bellow see the code generated by GraalVM EE.

##Before:
```
[info]   2.46%  │││││  │ ││    0x00007f039833640f: cmp    $0x30,%eax
[info]          │││││╭ │ ││    0x00007f0398336412: jl     0x00007f0398336473
[info]   3.68%  ││││││ │ ││    0x00007f0398336418: nopl   0x0(%rax,%rax,1)
[info]   2.50%  ││││││ │ ││    0x00007f0398336420: cmp    $0x3a,%eax
[info]          ││││││╭│ ││    0x00007f0398336423: jae    0x00007f0398336473  ;*ifeq {reexecute=0 rethrow=0 return_oop=0}
[info]          ││││││││ ││                                                  ; - com.github.plokhotnyuk.jsoniter_scala.core.JsonReader::parseLong@173 (line 1318)
[info]          ││││││││ ││                                                  ; - com.github.plokhotnyuk.jsoniter_scala.core.JsonReader::readLong@2 (line 376)
[info]          ││││││││ ││                                                  ; - com.github.plokhotnyuk.jsoniter_scala.benchmark.JsoniterScalaCodecs$$anon$27::d0@57 (line 63)
[info]          ││││││││ ││                                                  ; - com.github.plokhotnyuk.jsoniter_scala.benchmark.JsoniterScalaCodecs$$anon$27::decodeValue@3 (line 63)
[info]          ││││││││ ││                                                  ; - com.github.plokhotnyuk.jsoniter_scala.benchmark.JsoniterScalaCodecs$$anon$27::decodeValue@6 (line 63)
[info]          ││││││││ ││                                                  ; - com.github.plokhotnyuk.jsoniter_scala.core.JsonReader::read@46 (line 631)
[info]          ││││││││ ││                                                  ; - com.github.plokhotnyuk.jsoniter_scala.core.package$::readFromArray@15 (line 100)
[info]          ││││││││ ││                                                  ; - com.github.plokhotnyuk.jsoniter_scala.benchmark.ArrayOfLongsReading::jsoniterScala@19 (line 36)
[info]          ││││││││ ││                                                  ; - com.github.plokhotnyuk.jsoniter_scala.benchmark.generated.ArrayOfLongsReading_jsoniterScala_jmhTest::jsoniterScala_thrpt_jmhStub@1
```

##After:
```
[info]   0.29%  ││     │   0x00007f149c3342fa: lea    0x7fffffd0(%rsi),%edx  ;*iadd {reexecute=0 rethrow=0 return_oop=0}
[info]          ││     │                                                 ; - com.github.plokhotnyuk.jsoniter_scala.core.JsonReader::parseLong@159 (line 1318)
[info]          ││     │                                                 ; - com.github.plokhotnyuk.jsoniter_scala.core.JsonReader::readLong@2 (line 376)
[info]          ││     │                                                 ; - com.github.plokhotnyuk.jsoniter_scala.benchmark.JsoniterScalaCodecs$$anon$27::d0@57 (line 63)
[info]          ││     │                                                 ; - com.github.plokhotnyuk.jsoniter_scala.benchmark.JsoniterScalaCodecs$$anon$27::decodeValue@3 (line 63)
[info]          ││     │                                                 ; - com.github.plokhotnyuk.jsoniter_scala.benchmark.JsoniterScalaCodecs$$anon$27::decodeValue@6 (line 63)
[info]          ││     │                                                 ; - com.github.plokhotnyuk.jsoniter_scala.core.JsonReader::read@46 (line 631)
[info]          ││     │                                                 ; - com.github.plokhotnyuk.jsoniter_scala.core.package$::readFromArray@15 (line 100)
[info]          ││     │                                                 ; - com.github.plokhotnyuk.jsoniter_scala.benchmark.ArrayOfLongsReading::jsoniterScala@19 (line 36)
[info]          ││     │                                                 ; - com.github.plokhotnyuk.jsoniter_scala.benchmark.generated.ArrayOfLongsReading_jsoniterScala_jmhTest::jsoniterScala_thrpt_jmhStub@17 (line 119)
[info]   0.44%  ││     │   0x00007f149c334300: cmp    $0x8000000a,%edx
[info]          ││╭    │   0x00007f149c334306: jge    0x00007f149c334465  ;*ifeq {reexecute=0 rethrow=0 return_oop=0}
[info]          │││    │                                                 ; - com.github.plokhotnyuk.jsoniter_scala.core.JsonReader::parseLong@171 (line 1318)
[info]          │││    │                                                 ; - com.github.plokhotnyuk.jsoniter_scala.core.JsonReader::readLong@2 (line 376)
[info]          │││    │                                                 ; - com.github.plokhotnyuk.jsoniter_scala.benchmark.JsoniterScalaCodecs$$anon$27::d0@57 (line 63)
[info]          │││    │                                                 ; - com.github.plokhotnyuk.jsoniter_scala.benchmark.JsoniterScalaCodecs$$anon$27::decodeValue@3 (line 63)
[info]          │││    │                                                 ; - com.github.plokhotnyuk.jsoniter_scala.benchmark.JsoniterScalaCodecs$$anon$27::decodeValue@6 (line 63)
[info]          │││    │                                                 ; - com.github.plokhotnyuk.jsoniter_scala.core.JsonReader::read@46 (line 631)
[info]          │││    │                                                 ; - com.github.plokhotnyuk.jsoniter_scala.core.package$::readFromArray@15 (line 100)
[info]          │││    │                                                 ; - com.github.plokhotnyuk.jsoniter_scala.benchmark.ArrayOfLongsReading::jsoniterScala@19 (line 36)
[info]          │││    │                                                 ; - com.github.plokhotnyuk.jsoniter_scala.benchmark.generated.ArrayOfLongsReading_jsoniterScala_jmhTest::jsoniterScala_thrpt_jmhStub@17 (line 119)
```

**Comment by dougxc** [comment]
Tracked internally with GR-23729.

**Comment by dougxc** [comment]
@plokhotnyuk what are you comparing with `Before` and `After` in your example?

**Comment by plokhotnyuk** [comment]
@dougxc I have replaced captions to corresponding expressions in the comment above.

**Comment by dougxc** [comment]
Thanks. We will investigate further what it takes to apply this optimization.

**Comment by plokhotnyuk** [comment]
@dougxc Thank you a lot!

HotSpot JIT compiler already do this kind of optimization since JDK 8, but using `mov` + `sub` operations instead of `lea`:
```
[info]   0.06%    │     │ ││││  0x00007fea202ce92d: mov    %ecx,%edi
[info]   1.61%    │     │ ││││  0x00007fea202ce92f: add    $0xffffffd0,%edi
[info]   0.90%    │     │ ││││  0x00007fea202ce932: cmp    $0xa,%edi
[info]            │     │ ╰│││  0x00007fea202ce935: jae    0x00007fea202ce6a9  ;*if_icmpgt
[info]            │     │  │││                                                ; - com.github.plokhotnyuk.jsoniter_scala.core.JsonReader::parseLong@165 (line 1318)
[info]            │     │  │││                                                ; - com.github.plokhotnyuk.jsoniter_scala.core.JsonReader::readLong@2 (line 376)
[info]            │     │  │││                                                ; - com.github.plokhotnyuk.jsoniter_scala.benchmark.JsoniterScalaCodecs$$anon$27::d0@57 (line 63)
```

**Comment by lic9** [comment]
@dougxc @tkrodriguez **[canonicalizeConditionalCascade](https://github.com/oracle/graal/blob/master/compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/IfNode.java#L1108)**  might also get affected.

**Comment by tkrodriguez** [comment]
Yes the other cases are explicitly safe since they treat the null as meaning unknown.  Fixed in 2dfa6ab92a02f4c018376190ced992be9dd066e3
